### ISSUE TITLE
ingress-nginx-controller fails to start on SLES 15 SP2

### ISSUE THREAD

**Issue by rancher-max** [k3s-io/k3s#2778]
**Environmental Info:**
K3s Version: 
`k3s version v1.19.5+k3s1 (b11612e2)` (and should fail on other versions but did not validate to be sure)
RKE2 Version:
v1.18.13+rke2r1 and v1.19.5+rke2r1

Node(s) CPU architecture, OS, and Version: 

`Linux ip-172-31-24-244 5.3.18-24.37-default  #1  SMP Wed Nov 4 09:38:41 UTC 2020 (c145e08) x86_64 x86_64 x86_64 GNU/Linux`
Created in AWS using ami: `ami-0f052119b3c7e61d1`
```
# cat /etc/os-release
NAME="SLES"
VERSION="15-SP2"
VERSION_ID="15.2"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP2"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp2"
VARIANT_ID="sles-sap"
```

Cluster Configuration:

Single node

**Describe the bug:**

Exactly the same issue as described in https://github.com/kubernetes/ingress-nginx/issues/5991. However, I'm not convinced this is solely an AppArmor issue as disabling it (including ensuring it does not disable on startup and rebooting the node entirely) does not seem to resolve the issue. Also tried to do this in a fresh setup before installing either k3s or rke2 and saw the same results.

**Steps To Reproduce:**

K3s steps:
1. Install k3s
2. Attempt to install ingress-nginx-controller: `kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v0.34.1/deploy/static/provider/aws/deploy.yaml`
3. Notice`ingress-nginx-controller` pod stuck on `0/1 Running`

Rke2 steps:
1. Install rke2
2. Notice `rke2-ingress-nginx-controller` pod stuck at `0/1 Running`

**Expected behavior:**

pod should come up and be `1/1 Running` with no restarts and no failures in its logs.

**Actual behavior:**

See above

**Additional context / logs:**

Refer to linked issue

**Comment by rancher-max** [comment]
This appears to be an issue more with AWS and Azure provisioned instances. Booting the openSUSE image in AWS also has this same failure, but works in both Linode and on VirtualBox. The SLES 15 image ISO also succeeds on VirtualBox.

**Comment by davidnuzik** [comment]
If I had to take a guess here this will probably be a documentation effort. Like if those cloud images are missing some needed kernel modules or something? We'll just need to call it out in docs.
I think this is something we should understand and document as needed at a minimum. Plenty of other folks will be on AWS and Azure. This should work for them or be documented (what needs to be done to make such cloud images work).

**Comment by davidnuzik** [comment]
This issue blocks us claiming SLES support for K3s and RKE2 for Rancher 2.5.5 support matrix.

**Comment by davidnuzik** [comment]
More investigation is required. This is being moved to a mid-feb release.

**Comment by cseader** [comment]
I've set this up in my own lab with SLES 15 SP2 I have updated all SLES 15 SP2 hosts with the latest updates as of Jan. 18th 2021.

Error in logs:
rke2-ingress-nginx-controller 2021/01/09 00:24:53 [notice] 340#340: signal process started                                                                                                                      
rke2-ingress-nginx-controller 2021/01/09 00:24:53 [alert] 340#340: kill(25, 1) failed (13: Permission denied)                                                                                                   
rke2-ingress-nginx-controller nginx: [alert] kill(25, 1) failed (13: Permission denied)                                                                                                                         
rke2-ingress-nginx-controller W0109 00:24:53.893819       8 queue.go:130] requeuing kube-system/rke2-ingress-nginx-default-backend, err exit status 1                                                          
rke2-ingress-nginx-controller 2021/01/09 00:24:53 [notice] 340#340: signal process started                                                                                                                      
rke2-ingress-nginx-controller 2021/01/09 00:24:53 [alert] 340#340: kill(25, 1) failed (13: Permission denied)                                                                                                   
rke2-ingress-nginx-controller nginx: [alert] kill(25, 1) failed (13: Permission denied)                                                                                                                         
rke2-ingress-nginx-controller I0109 00:24:57.144680       8 controller.go:137] Configuration changes detected, backend reload required.                                                                         
rke2-ingress-nginx-controller E0109 00:24:57.216879       8 controller.go:149] Unexpected failure reloading the backend: 

going through kernel  module requirements right now.

**Comment by cseader** [comment]
For some reason we have a port conflict 
2021-01-08T21:13:06.26767024-07:00 stderr F I0109 04:13:06.267445       7 flags.go:215] Watching for Ingress class: nginx
2021-01-08T21:13:06.267715292-07:00 stdout F -------------------------------------------------------------------------------
2021-01-08T21:13:06.267758091-07:00 stdout F NGINX Ingress controller
2021-01-08T21:13:06.267766239-07:00 stdout F   Release:       0.30.0
2021-01-08T21:13:06.267771178-07:00 stdout F   Build:         git-fbdaf0775
2021-01-08T21:13:06.26777675-07:00 stdout F   Repository:    https://github.com/rancher/ingress-nginx.git
2021-01-08T21:13:06.267782683-07:00 stdout F   nginx version: nginx/1.17.8
2021-01-08T21:13:06.267785738-07:00 stdout F 
2021-01-08T21:13:06.267790169-07:00 stdout F -------------------------------------------------------------------------------
2021-01-08T21:13:06.267793073-07:00 stdout F 
2021-01-08T21:13:06.268917807-07:00 stderr F F0109 04:13:06.268676       7 main.go:65] port 10245 is already in use. Please check the flag --profiler-port

The port shown is the default port that Ingress controller uses.

**Comment by nikkelma** [comment]
Inspected the OS audit logs and found lines like ...
```
type=AVC msg=audit(1611173469.740:321): apparmor="DENIED" operation="signal" profile="cri-containerd.apparmor.d" pid=2090 comm="kube-apiserver" requested_mask="receive" denied_mask="receive" signal=urg peer="cri-containerd.apparmor.d"
...
type=AVC msg=audit(1611173236.769:727): apparmor="DENIED" operation="signal" profile="cri-containerd.apparmor.d" pid=5686 comm="nginx-ingress-c" requested_mask="receive" denied_mask="receive" signal=pipe peer="cri-containerd.apparmor.d"
```

AppArmor is clearly blocking the signals, but supposedly AppArmor profiles aren't trying to be used at all currently in k3s/RKE2. This hints that there could be larger issues in play here since both the API server and the ingress controller were being blocked from killing other container processes.

Worked around the issue with a `/var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl` containing the following. Compare to [original file](https://github.com/k3s-io/k3s/blob/3ea696815b10613b287b87a8d10b8e26e7d23d9c/pkg/agent/templates/templates.go#L17-L74) ... differences are highlighted below, too.

```
[plugins.opt]
  path = "{{ .NodeConfig.Containerd.Opt }}"
[plugins.cri]
  stream_server_address = "127.0.0.1"
  stream_server_port = "10010"
  enable_selinux = {{ .NodeConfig.SELinux }}
{{- if .IsRunningInUserNS }}
  disable_cgroup = true
  # ---
  # remove duplicated property
  # disable_apparmor = true
  # ---
  restrict_oom_score_adj = true
{{end}}
  # ---
  # disable apparmor always instead of only when running in user namespace
  disable_apparmor = true
  # ---
{{- if .NodeConfig.AgentConfig.PauseImage }}
  sandbox_image = "{{ .NodeConfig.AgentConfig.PauseImage }}"
{{end}}
{{- if .NodeConfig.AgentConfig.Snapshotter }}
[plugins.cri.containerd]
  disable_snapshot_annotations = true
  snapshotter = "{{ .NodeConfig.AgentConfig.Snapshotter }}"
{{end}}
{{- if not .NodeConfig.NoFlannel }}
[plugins.cri.cni]
  bin_dir = "{{ .NodeConfig.AgentConfig.CNIBinDir }}"
  conf_dir = "{{ .NodeConfig.AgentConfig.CNIConfDir }}"
{{end}}
[plugins.cri.containerd.runtimes.runc]
  runtime_type = "io.containerd.runc.v2"
{{ if .PrivateRegistryConfig }}
{{ if .PrivateRegistryConfig.Mirrors }}
[plugins.cri.registry.mirrors]{{end}}
{{range $k, $v := .PrivateRegistryConfig.Mirrors }}
[plugins.cri.registry.mirrors."{{$k}}"]
  endpoint = [{{range $i, $j := $v.Endpoints}}{{if $i}}, {{end}}{{printf "%q" .}}{{end}}]
{{end}}
{{range $k, $v := .PrivateRegistryConfig.Configs }}
{{ if $v.Auth }}
[plugins.cri.registry.configs."{{$k}}".auth]
  {{ if $v.Auth.Username }}username = {{ printf "%q" $v.Auth.Username }}{{end}}
  {{ if $v.Auth.Password }}password = {{ printf "%q" $v.Auth.Password }}{{end}}
  {{ if $v.Auth.Auth }}auth = {{ printf "%q" $v.Auth.Auth }}{{end}}
  {{ if $v.Auth.IdentityToken }}identitytoken = {{ printf "%q" $v.Auth.IdentityToken }}{{end}}
{{end}}
{{ if $v.TLS }}
[plugins.cri.registry.configs."{{$k}}".tls]
  {{ if $v.TLS.CAFile }}ca_file = "{{ $v.TLS.CAFile }}"{{end}}
  {{ if $v.TLS.CertFile }}cert_file = "{{ $v.TLS.CertFile }}"{{end}}
  {{ if $v.TLS.KeyFile }}key_file = "{{ $v.TLS.KeyFile }}"{{end}}
  {{ if $v.TLS.InsecureSkipVerify }}insecure_skip_verify = true{{end}}
{{end}}
{{end}}
{{end}}
```

**Comment by bashofmann** [comment]
> This hints that there could be worse symptoms if explored more deeply since the API server is also being blocked from killing other container processes.

This explains an issue I had the other day with k3s on sles where Pods were stuck in `Terminating`.

**Comment by brandond** [comment]
Technically the API server doesn't kill processes - the apiserver sets the deletedTimestamp on the pod, which triggers the kubelet to ask the runtime to terminate the pods, who in turn sends TERM/KILL signals to the processes in the pods. But yeah, if AppArmor is for some reason blocking signals everything is going to fall apart.

I'm not finding it now, but there was a discussion regarding the fact that the Suse cloud images set the polkit profile to restrictive instead of standard: https://github.com/openSUSE/polkit-default-privs/tree/master/profiles

I don't see anything in there that jumps out at me as locking down apparmor signaling, but I believe it is related.

**Comment by brandond** [comment]
@rancher-max can you try following the instructions at https://documentation.suse.com/sles/15-SP1/html/SLES-all/cha-security-policykit.html#sec-security-policykit-policies-default to set the profile to `standard` on one of the hosts that were failing? If this is indeed what's breaking RKE2 then we just need to document that step as a prerequisite.

**Comment by rancher-max** [comment]
### In RKE2, validated everything except nodeport functional using only the workaround mentioned in https://github.com/k3s-io/k3s/issues/2778#issuecomment-765056128
**Steps I took were:**
1. Create a fresh SLES 15 SP2 VM
2. Create containerd config:
```
sudo mkdir -p /var/lib/rancher/rke2/agent/etc/containerd/
sudo vi /var/lib/rancher/rke2/agent/etc/containerd/config.toml.tmpl
# Paste contents from the above mentioned comment
```
3. Create rke2 config:
```
sudo mkdir -p /etc/rancher/rke2 && \
sudo vi /etc/rancher/rke2/config.yaml
# Any config options desired. I used cloud-provider:
write-kubeconfig-mode: 644
tls-san: 
  - mydns.fake.com
cloud-provider-name:  "aws"
node-name: <result of $(hostname -f)>
token: "secret"
```
4. Run rke2:
```
curl -sfL https://get.rke2.io | sudo INSTALL_RKE2_VERSION=v1.19.5+rke2r1 sh - 
sudo systemctl enable rke2-server.service && \
sudo systemctl start rke2-server.service
```
**Results:**
- All system pods came up functionally
- Import into rancher - success
- Run preupgrade checks - success
- Upgrade to v1.19.7+rke2r1 - success
- Run postupgrade checks - success
- Run regression checks - Most passed. NodePort tests failed.

I am unsure exactly why nodeport wasn't functional in this case. Further investigation is needed, but my thought is I might need to do a mix of the workaround with apparmor in containerd combined with the polkit settings.
_All that said, we have the majority of functionality covered now with the workaround._
I will do more checks when I return from vacation to see if I can workaround the NodePort issue.

**Theory for why workaround is needed:**
The difference between the customized containerd config vs default is disabling apparmor in all cases, not just if it hits the if condition `if .IsRunningInUserNS`. Below are my thoughts on that:
```
Why do we do the check for IsRunningInUserNS? 
https://github.com/k3s-io/k3s/blob/2555ffe9f8c279bd75a6449a521385fc36d43e8f/vendor/github.com/containerd/containerd/sys/userns_linux.go#L33

I"m wondering if SLES for some reason hits the non-linux condition? https://github.com/k3s-io/k3s/blob/2555ffe9f8c279bd75a6449a521385fc36d43e8f/vendor/github.com/containerd/containerd/sys/userns_unsupported.go#L23

Otherwise, it should hit the condition and return true and disable apparmor based on my findings:
https://github.com/k3s-io/k3s/blob/2555ffe9f8c279bd75a6449a521385fc36d43e8f/vendor/github.com/containerd/containerd/sys/userns_linux.go#L56-L58
$ cat /proc/self/uid_map
         0          0 4294967295
```

**Comment by brandond** [comment]
It appears that the proper fix for this is in https://github.com/containerd/containerd/pull/4467

**Comment by rancher-max** [comment]
This has been validated now in both rke2 using the latest beta releases and in k3s using the commit ids from 1.18, 1.19, and 1.20 branches (ids, respectively: `8c9608b26f07e96a7445f46eeb341a90a33b68a7`, `e34ffdba565067bce8a9da5b79d226f2aad7de4c`, and `1d85a6a30a781ba8644c8f7e3ceec3cc0d21238b`).

Deploying the yaml mentioned in the issue correctly deploys ingress-nginx-controller now, but there is a port conflict on the svclb. Editing that daemonset to use different ports correctly deploys that as well. The next release of rke2 and k3s should work fine on SLES.

**Comment by davidnuzik** [comment]
Re-opening for the purposes of retesting with https://github.com/k3s-io/containerd/pull/14
This IS FIXED in v1.18.16 but we had a regression in v1.18.17-rc's so we had to resolve an issue in k3s-io/containerd 1.3 branch to fix this again.

**Comment by rancher-max** [comment]
This has been resolved in v1.18.17-rc2+k3s1. Leaving this open to validate the rke2 fix as well.

**Comment by rancher-max** [comment]
Validated in v1.18.17-rc2+rke2r1 as well.
